### PR TITLE
Status reporting

### PR DESCRIPTION
**Pull request by bboozzoo**
First take on status reporting.

Sending status over HTTP is captured in `StatusClient`, implemented in the first part of this series. Failure to send a status report results in an error that the caller is expected to handle in some way.

`Controller` interface got a new function `ReportStatusUpdate` that returns a `menderError` or nil.

The fun part is captured in state handling. Basically each 'update' state (commit, install, fetch, reboot) can transition to a new error state - `UpdateErrorState`. Also each of these states carries the update information. This is nice, because we don't need to keep it as part of Controller's internal state. Update information should be saved as soon as we decide to apply an update (ideally when update check returns valid update information) and restored right after (returned from?) `HasUpgrade()`.

@kacf @pasinskim

### CODE REVIEW

File: client_status.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		return errors.Wrapf(err, "failed to prepare status report request")
+	}
+
+	r, err := api.Do(req)
+	if err != nil {
+		log.Error("failed to report status: ", err)
+		return errors.Wrapf(err, "reporting status failed")
+	}
+
+	if r.StatusCode != http.StatusNoContent {
```
Comment (kacf) [inline client_status.go]: What is `StatusNoContent`? Is that the expected response?

File: client_status.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	out := &bytes.Buffer{}
+	enc := json.NewEncoder(out)
+	enc.Encode(&report)
+
+	req, err := http.NewRequest(http.MethodPut, url, out)
+	if err != nil {
+		return nil, err
+	}
+	return req, nil
```
Comment (kacf) [inline client_status.go]: I think you could shorten this to simply return `http.NewRequest` directly, which is (almost) equivalent (`req` isn't guaranteed to be `nil` in case of error, but this is unlikely to be significant).

File: client_status.go (hunk 1 above)
Comment (bboozzoo) [inline client_status.go]: Yes, the API proposal lists 204 (No Content) as response.

**Comment by kacf** [comment]
Looks good to me otherwise!
